apps/vapi/RenderManager.cpp

```diff
@@ -0,0 +1,239 @@
+        width = vParams->GetValueLong(ViewpointParams::CustomFramebufferWidthTag, 0);
+        height = vParams->GetValueLong(ViewpointParams::CustomFramebufferHeightTag, 0);
+        if (width == 0) width = 1;
+        if (height == 0) height = 1;
+
+        int maxSize;
+        glGetIntegerv(GL_MAX_TEXTURE_SIZE, &maxSize);
+        if (width > maxSize) {
+            width = maxSize;
+            vParams->SetValueLong(ViewpointParams::CustomFramebufferWidthTag, ViewpointParams::CustomFramebufferWidthTag, width);
+            LogFatal("Selected width is larger than your OpenGL implementation supports");
+        }
+        if (height > maxSize) {
+            height = maxSize;
+            vParams->SetValueLong(ViewpointParams::CustomFramebufferHeightTag, ViewpointParams::CustomFramebufferHeightTag, height);
+            LogFatal("Selected height is larger than your OpenGL implementation supports");
+        }
+
+        float fa = width / (float)height;
+        float wa = wWidth / (float)wHeight;
+
+        if (fa >= wa) {
+            int x = 0;
+            int y = (wHeight / 2) - (wHeight / fa * wa / 2);
+            int w = wWidth;
+            int h = wHeight / fa * wa;
+            glViewport(x, y, w, h);
+        } else {
+            int x = (wWidth / 2) - (wWidth * fa / wa / 2);
+            int y = 0;
+            int w = wWidth * fa / wa;
+            int h = wHeight;
+            glViewport(x, y, w, h);
+        }
+    } else {
+        glViewport(0, 0, width, height);
+    }
+
+    mm->MatrixModeProjection();
+    mm->LoadIdentity();
+
+    float w = (float)width / (float)height;
+
+    if (vParams->GetProjectionType() == ViewpointParams::MapOrthographic) {
+        Trackball trackball;
+        double    center[3];
+        vParams->GetRotationCenter(center);
+        trackball.setFromFrame(posvec, dirvec, upvec, center, true);
+        float s = trackball.GetOrthoSize();
+        mm->Ortho(-s * w, s * w, -s, s, nearDist, farDist);
+        vParams->SetOrthoProjectionSize(trackball.GetOrthoSize());
+    } else {
+        double fov = vParams->GetFOV();
+        mm->Perspective(glm::radians(fov), w, nearDist, farDist);
+    }
+
+    double pMatrix[16];
+    mm->GetDoublev(MatrixManager::Mode::Projection, pMatrix);
+
+    bool enabled = _controlExec->GetSaveStateEnabled();
+    _controlExec->SetSaveStateEnabled(false);
+
+    vParams->SetProjectionMatrix(pMatrix);
+
+    _controlExec->SetSaveStateEnabled(enabled);
+
+    mm->MatrixModeModelView();
+}
+
+void RenderManager::setUpModelViewMatrix()
+{
+    ParamsMgr *      paramsMgr = _controlExec->GetParamsMgr();
+    ViewpointParams *vParams = paramsMgr->GetViewpointParams(getWinName());
+
+    double m[16];
+    vParams->GetModelViewMatrix(m);
+    _glManager->matrixManager->LoadMatrixd(m);
+}
+
+
+
+int RenderManager::Render(String imagePath)
+{
+    //    GL_ERR_BREAK();
+    if (!_glManager) _glManager = new GLManager;
+
+    static bool temp = false;
+    if (!temp) {
```

Comment:
Can we use a member variable here?

Reply:
This is some temporary code so the demo works for the WIP and there is a TODO reminder to remove it 2 lines down. It will be removed later.